more_itertools/more.py

```diff
@@ -2306,3 +2307,30 @@ def replace(iterable, pred, substitutes, count=None, window_size=1):
         # yield the first item from the window.
         if w and (w[0] is not _marker):
             yield w[0]
+
+
+def _is_collection(iterable):
+    return hasattr(iterable, '__getitem__') and hasattr(iterable, '__len__')
+
+
+def partitions(iterable):
+    """Yield all possible partitions of *iterable*.
+    Partition x is defined so that:
+    1. list(x) = [x_0, ..., x_n]
+    2. x_0 + ... + x_n = list(iterable)
+    3. all(isinstance(x_i, (type(iterable), list)) for x_i in x) == True
+    4. 1 <= n <= ilen(iterable)
+    5. isinstance(x, list) == True
+
+        >>> iterable = 'abcd'
+        >>> list(partitions(iterable))
+        [['abcd'], ['a', 'bcd'], ['ab', 'cd'], ['abc', 'd'], \
+['a', 'b', 'cd'], ['a', 'bc', 'd'], ['ab', 'c', 'd'], ['a', 'b', 'c', 'd']]
+
+    """
+    collection = iterable if _is_collection(iterable) else list(iterable)
```

Comment:
I think I'd prefer if this was just `list(iterable)`.

 reason: none of the other functions in this module . For example `list(substrings(iterable))` gives `[('a',), ('b',), ('a', 'b')]` - not `['a', 'b', 'ab']`.

Second reason: `_is_collection` is checking for a sequence, which is [kind of messy](https://stackoverflow.com/questions/43566044/what-is-pythons-sequence-protocol). It passes for `dict()` objects, using the keys. But it fails for `set()` objects. Is there a principled reason why we should allow one but not the other, for versions below 3.7?

(In [`SequenceView`](https://github.com/erikrose/more-itertools/blob/master/more_itertools/more.py#L1918) we just check to see if the object inherits from `Sequence`.)

Reply:
> First reason: For example list(substrings(iterable)) gives [('a',), ('b',), ('a', 'b')] - not ['a', 'b', 'ab'].

And that's a bad thing in my opinion. Why the function which is called `substrings` returns generator (good) of tuples of strings with length 1? I may ask another way around. Do you think that `('a', 'b')` is substring of `'abc'`. I don't think so. Read for example documentation of `str.find`: "Return the lowest index in S where substring sub is found". OK, now I will try to put substring in the sense `more-itertools`:
```python
In [1]: 'abc'.find(('a', 'b'))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-1-fabec44c37f8> in <module>
----> 1 'abc'.find(('a', 'b'))

TypeError: must be str, not tuple
```
Oops. That doesn't work.

> Second reason: _is_collection is checking for a sequence, which is kind of messy. It passes for dict() objects, using the keys. But it fails for set() objects. Is there a principled reason why we should allow one but not the other, for versions below 3.7?

I'm checking for [Collection](https://docs.python.org/3/library/collections.abc.html#collections.abc.Collection). That's not a `Sequence`. But you are right, I missed the dict. What do you think of this code:
```python
def _is_collection(iterable):
    try:
        iterable[slice()]
    except TypeError:
        return False
    return hasattr(iterable, '__len__')
```

Reply:
I've changed the `_is_collection`.

Reply:
I can understand why you want `str`s together, but the rest of the library doesn't do that. So if we can do `list` on the iterable, then I'll be happy to add this.

Reply:
> So, I propose to accept this here, and change it everywhere else to make library behave more predictable.

This library is predictable in that it extends `itertools`, which generally returns iterators.  The iterator is an abstract artifact that encompasses all iterables, including strings.  Changing the library to return strings limits it.  Consider the dependents on this API.  That change would cripple a lot of code.

> Why the function which is called substrings returns generator (good) of tuples of strings with length 1 (very strange)?

Because it handles more than strings, e.g. `list(substring([1, 2, 3]))`.  Sure, empty items is weird, but the [definition](https://en.wikipedia.org/wiki/Substring) predates the library; see also [`subsequence`](https://en.wikipedia.org/wiki/Subsequence) and Haskell's implementation of substring.  Also note, the itertools `powerset` recipe returns an iterator, not a set.

Reply:
Well, obviously, I'm outvoted, so, even though I still do not agree, I'm changing function to always return generator of list of lists.